File: src/lib/consistency.ts
import { format, subDays } from "date-fns";
import {
  getPeriodKey,
  isHabitDueOn,
  type HabitSchedule,
} from "@/lib/schedule";

export const CONSISTENCY_WINDOW_DAYS = 14;
export const STREAK_THRESHOLD = 1;

export function toDateKey(date: Date = new Date()): string {
  return format(date, "yyyy-MM-dd");
}

export function dateKeysForWindow(
  days: number = CONSISTENCY_WINDOW_DAYS,
  end: Date = new Date(),
): string[] {
  return Array.from({ length: days }, (_, index) =>
    toDateKey(subDays(end, days - 1 - index)),
  );
}

function parseDateKey(key: string) {
  return new Date(`${key}T12:00:00`);
}

export type HabitForScore = HabitSchedule & {
  id: string;
  createdAt?: Date | null;
};

export type CompletionForScore = {
  habitId: string;
  completedOn: string;
  slot?: number;
};

function habitRatioOnDay(
  habit: HabitForScore,
  dayKey: string,
  completions: CompletionForScore[],
) {
  const day = parseDateKey(dayKey);
  if (!isHabitDueOn(habit, day, habit.createdAt ?? null)) {
    return null;
  }

  const periodKey = getPeriodKey(habit, day, habit.createdAt ?? null);
  const times = Math.max(1, habit.timesPerPeriod);
  const slots = new Set(
    completions
      .filter(
        (completion) =>
          completion.habitId === habit.id &&
          completion.completedOn === periodKey,
      )
      .map((completion) => completion.slot ?? 0),
  );

  return Math.min(1, slots.size / times);
}

export function computeConsistencyScore(params: {
  habits: HabitForScore[];
  completions: CompletionForScore[];
  windowDays?: number;
  end?: Date;
}): number {
  const {
    habits,
    completions,
    windowDays = CONSISTENCY_WINDOW_DAYS,
    end = new Date(),
  } = params;

  if (habits.length === 0) {
    return 0;
  }

  const keys = dateKeysForWindow(windowDays, end);
  let total = 0;
  let counted = 0;

  for (const key of keys) {
    const ratios = habits
      .map((habit) => habitRatioOnDay(habit, key, completions))
      .filter((value): value is number => value !== null);

    if (ratios.length === 0) continue;
    total += ratios.reduce((sum, value) => sum + value, 0) / ratios.length;
    counted += 1;
  }

  return counted === 0 ? 0 : total / counted;
}

export function computeStreak(params: {
  habits: HabitForScore[];
  completions: CompletionForScore[];
  threshold?: number;
  end?: Date;
}): number {
  const {
    habits,
    completions,
    threshold = STREAK_THRESHOLD,
    end = new Date(),
  } = params;

  if (habits.length === 0) {
    return 0;
  }

  let streak = 0;
  for (let offset = 0; offset < 365; offset += 1) {
    const key = toDateKey(subDays(end, offset));
    const ratios = habits
      .map((habit) => habitRatioOnDay(habit, key, completions))
      .filter((value): value is number => value !== null);

    if (ratios.length === 0) {
      if (offset === 0) continue;
      break;
    }

    const dayScore =
      ratios.reduce((sum, value) => sum + value, 0) / ratios.length;

    if (dayScore >= threshold) {
      streak += 1;
    } else if (offset === 0) {
      continue;
    } else {
      break;
    }
  }

  return streak;
}

File: apps/api/app/domain/consistency.py
"""14-day consistency score + streak, ported from src/lib/consistency.ts."""

from datetime import date, datetime, timedelta

from app.domain.schedule import get_period_key, is_habit_due_on

CONSISTENCY_WINDOW = 14
CALENDAR_WINDOW = 28
STREAK_THRESHOLD = 1.0


def _date_keys(days: int, end: date | None = None) -> list[str]:
    end = end or date.today()
    return [(end - timedelta(days=days - 1 - i)).isoformat() for i in range(days)]


def _habit_ratio(
    habit: dict,
    day_key: str,
    completions: list[dict],
) -> float | None:
    d = date.fromisoformat(day_key)
    freq = habit["frequency"]
    custom = habit.get("custom_every_days")
    created = habit.get("created_at")

    if not is_habit_due_on(freq, custom, d, created):
        return None

    pk = get_period_key(freq, habit.get("times_per_period", 1), custom, d, created)
    times = max(1, habit.get("times_per_period", 1))
    slots = {
        c["slot"]
        for c in completions
        if c["habit_id"] == habit["id"] and c["completed_on"] == pk
    }
    return min(1.0, len(slots) / times)


def compute_consistency(
    habits: list[dict],
    completions: list[dict],
    window: int = CONSISTENCY_WINDOW,
    end: date | None = None,
) -> float:
    if not habits:
        return 0.0
    keys = _date_keys(window, end)
    total = counted = 0.0
    for key in keys:
        ratios = [r for h in habits if (r := _habit_ratio(h, key, completions)) is not None]
        if not ratios:
            continue
        total += sum(ratios) / len(ratios)
        counted += 1
    return total / counted if counted else 0.0


def compute_streak(
    habits: list[dict],
    completions: list[dict],
    threshold: float = STREAK_THRESHOLD,
    end: date | None = None,
) -> int:
    if not habits:
        return 0
    end = end or date.today()
    streak = 0
    for offset in range(365):
        key = (end - timedelta(days=offset)).isoformat()
        ratios = [r for h in habits if (r := _habit_ratio(h, key, completions)) is not None]
        if not ratios:
            if offset == 0:
                continue
            break
        day_score = sum(ratios) / len(ratios)
        if day_score >= threshold:
            streak += 1
        elif offset == 0:
            continue
        else:
            break
    return streak


def compute_daily_scores(
    habits: list[dict],
    completions: list[dict],
    window: int = CALENDAR_WINDOW,
    end: date | None = None,
) -> list[dict]:
    keys = _date_keys(window, end)
    result = []
    for key in keys:
        if not habits:
            result.append({"date": key, "score": None})
            continue
        ratios = [r for h in habits if (r := _habit_ratio(h, key, completions)) is not None]
        if not ratios:
            result.append({"date": key, "score": None})
        else:
            result.append({"date": key, "score": sum(ratios) / len(ratios)})
    return result
